server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/ClusterAllocationSimulationTests.java

```diff
@@ -0,0 +1,563 @@
+        @Override
+        public Metadata.Builder withWriteLoadForecastForWriteIndex(String dataStreamName, Metadata.Builder metadata) {
+            throw new AssertionError("should not be rolling over in simulation");
+        }
+
+        @Override
+        @SuppressForbidden(reason = "This is simulation implementation")
+        public OptionalDouble getForecastedWriteLoad(IndexMetadata indexMetadata) {
+            return indexMetadata.getForecastedWriteLoad();
+        }
+    };
+
+    public void testBalanceQuality() throws IOException {
+
+        final var shardSizesByIndex = new HashMap<String, Long>();
+        final var tiers = new String[] { DataTier.DATA_HOT, DataTier.DATA_WARM, DataTier.DATA_COLD };
+        final var tierSizes = Arrays.stream(tiers).collect(Collectors.toMap(Function.identity(), s -> 0L));
+        final var tierWriteLoads = Arrays.stream(tiers).collect(Collectors.toMap(Function.identity(), s -> 0.0));
+        final var maxSizeVariance = ByteSizeValue.ofMb(20).getBytes();
+
+        final var metadataBuilder = Metadata.builder();
+        for (var dataStreamIndex = between(10, 50); dataStreamIndex >= 0; dataStreamIndex -= 1) {
+            final var hotIndices = between(2, 20);
+            final var warmIndices = between(10, 30);
+            final var coldIndices = between(10, 100);
+            final var hotShards = randomFrom(1, 2, 4, 8);
+            final var shrunkShards = randomFrom(1, 2);
+            final var approxIndexSize = randomFrom(ByteSizeValue.ofGb(1), ByteSizeValue.ofGb(10), ByteSizeValue.ofGb(50));
+
+            final var indexCount = hotIndices + warmIndices + coldIndices;
+            for (var index = 0; index < indexCount; index++) {
+                final var tier = index < coldIndices ? DataTier.DATA_COLD
+                    : index < coldIndices + warmIndices ? DataTier.DATA_WARM
+                    : DataTier.DATA_HOT;
+                final var indexName = Strings.format("index-%03d-%s-%03d", dataStreamIndex, tier.charAt(5), index);
+                final var shardCount = DataTier.DATA_HOT.equals(tier) ? hotShards : shrunkShards;
+                final var replicaCount = DataTier.DATA_COLD.equals(tier) ? 0 : 1;
+                final var indexWriteLoad = index == indexCount - 1 ? (scaledRandomIntBetween(1, 8000) / 1000.0) : 0.0;
+
+                metadataBuilder.put(
+                    IndexMetadata.builder(indexName)
+                        .settings(
+                            Settings.builder()
+                                .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, shardCount)
+                                .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, replicaCount)
+                                .put(IndexMetadata.SETTING_VERSION_CREATED, Version.CURRENT)
+                                .put(IndexMetadata.INDEX_ROUTING_REQUIRE_GROUP_PREFIX + ".fake_tier", tier)
+                        )
+                        .indexWriteLoadForecast(indexWriteLoad)
```

Comment:
Can we add the forecasted disk usage here too?

Reply:
Yep, see 9c81fdf2ed4164c560ad37c0a648267ca230f813.